execution_workflows/APAtrap/modules/identify_distal_3utr.nf

```diff
@@ -0,0 +1,67 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+def modules = params.modules.clone()
+def inputs    = modules['identify_distal_3utr']
+def workflow_option = params.workflow.clone()
+def run_differential = workflow_option['run_differential']
+
+/*
+    Run the first step of APAtrap: identifyDistal3UTR to refine
+    annotated 3'UTRs and identify novel 3'UTRs or 3'UTR extensions.
+*/
+process IDENTIFY_DISTAL_3UTR {
+
+        tag "$sample"
+        publishDir "${params.outdir}/apatrap", mode: params.publish_dir_mode
+        container "docker.io/apaeval/apatrap:latest"
+
+        input:
+        val sample_bedgraph_files_dir
+        tuple path(genome_file), val(sample), path(utr_input_bedgraph)
+
+        output:
+        tuple val(sample), path(utr_output), emit: ch_predictapa_input
+
+        script:
+        window_size = inputs.w
+        extension_size = inputs.e
+        min_coverage = inputs.c
+        min_percentage = inputs.p
+        pwd = "$PWD/${params.outdir}/$sample_bedgraph_files_dir"
+        // if run differential, all sample files have to be ran together
+        if (run_differential) {
+            utr_output = "predictapa_input.bed"
+            """
+            #!/bin/bash
+            sample_files=""
+            for file in "$pwd"/*
+            do
+                echo \$file
+                sample_files+="\$file "
+            done
+            identifyDistal3UTR -i \$sample_files \
+                -m $genome_file \
+                -o $utr_output \
+                -w $window_size \
+                -e $extension_size \
+                -c $min_coverage \
+                -p $min_percentage
+            """
```

Comment:
No clue, but maybe something like this: https://nextflow-io.github.io/patterns/index.html#_process_per_file_output

Reply:
@SamBryce-Smith @ninsch3000 Are you guys referring do line 40-44 where we're concatenating the list of samples? The reason I'm doing this is exactly because nextflow processes things per file output (that CJ linked), but we don't want this since we need all the file outputs/sample files together in one identifyDistal3UTR call. Hence the bash script file name concatenation 😆  So yeah I'm not sure if nextflow can do this (couldn't find anything online) since it's innately designed to process per file output

Reply:
Yep line 40-44! That's fair enough, it works perfectly fine so it's not a sticking point for me. 

I'm surprised there isn't a way to make a Nextflow rule 'wait' for all per-sample pre-processing to be done/have a rule that takes all samples as input though. Unless @yuukiiwa or another Nextflow expert wants to chime in XD